app/models/user.rb

```diff
@@ -121,6 +122,46 @@ def owns_project?(project_id)
     projects.exists?(project_id)
   end
 
+  def get_user_preference(key)
+    user_preference = user_preferences.find_by(key: key)
+
+    return user_preference.value unless user_preference.nil?
+
+    return UserPreference::USER_PREFERENCE_METADATA[key][:default]
+  end
+
+  def set_user_preference(key, value)
+    user_preference = user_preferences.find_or_initialize_by(key: key)
+
+    user_preference.value = value
+    user_preference.save!
+  end
+
+  # Remove any user preferences gated on allowed_features that the user doesn't have access to.
+  def viewable_user_preference_keys
+    parsed_allowed_feature_list = allowed_feature_list
+    UserPreference::USER_PREFERENCE_METADATA.select do |_key, metadata|
+      metadata[:allowed_feature].nil? || parsed_allowed_feature_list.include?(metadata[:allowed_feature])
```

Comment:
maybe `metadata[:allowed_feature].present?` in case people accidentally use `allowed_feature: ""` ?